packages/ai-jsx/src/batteries/constrained-output.tsx

```diff
@@ -257,3 +267,39 @@ async function* ObjectCompletionWithRetry(
     }
   );
 }
+
+/**
+ * A {@link ChatCompletion} component that constrains the output to be a valid JSON string.
+ * It (ab)uses OpenAI function calls to generate the JSON string.
+ *
+ * @returns A string that is a valid JSON or throws an error.
+ */
+export async function* JsonChatCompletionFunctionCall(
+  { schema, children, ...props }: ModelPropsWithChildren & { schema: z.Schema },
+  { render }: AI.ComponentContext
+) {
+  const childrenWithCompletion = (
+    <ChatCompletion
+      experimental_streamFunctionCallOnly
+      {...props}
+      functionDefinitions={{
+        print: {
+          description: 'Prints the response in a human readable format.',
+          parameters: schema,
+        },
+      }}
+    >
+      {children}
+      <SystemMessage>
+        Your response must use the `print` function that is provided. No other explanation needed. do not respond with
+        an assistant message. Just call the function.
+      </SystemMessage>
+    </ChatCompletion>
+  );
+
+  const frames = render(childrenWithCompletion);
+  for await (const frame of frames) {
+    yield JSON.stringify(JSON.parse(frame).arguments);
+  }
+  return JSON.stringify(JSON.parse(await frames).arguments);
```

Comment:
@petersalas, is this the right pattern to modify the result of an inner component as it's being passed up the render tree? Or am I just supposed to use the `map` arg here?

Reply:
Either works

Reply:
Hmm. I switched to this:

```tsx
  return render(childrenWithCompletion, {
    map: (frame) => JSON.stringify(JSON.parse(frame).arguments)
  });
```

And it doesn't apply the mapping function to the final result. That seems consistent with the other places we call `render`, but it also seems like it'll be a footgun to people with less context.

I then tried a few ways to also act on the final result, but it didn't work – what's the recommended approach here?

Reply:
Perhaps relatedly, something odd is going on here.

When I run `yarn workspace examples demo:validated-generation`, this works fine as-is.

However, when I rebase https://github.com/fixie-ai/ai-jsx/pull/186 to be on top of this branch and run it, the render result emitted from `JsonChatCompletionFunctionCall` looks good, but the render result of `JsonChatCompletion` is a doubly-quoted string: `"\"{a: 123}\""`. This means that when the caller calls `JSON.parse`, it gets a string, not an object.

I'm confused as to why the output of `JCCFC` looks good and the output of `JCC` looks bad, given that this is all `JCC` does:

```
      return yield* render(<JsonChatCompletionFunctionCall schema={schema} {...props} />);
```

Reply:
```ts
return map(yield* render(..., { map }));
```

```ts
const iterator = render(...)[Symbol.asyncIterator]();
while (true) {
    const next = await iterator.next();
    const value = map(next.value);
    if (next.done) {
        return value;
    } else {
        yield value;
    }
}
```

```ts
const rendered = render(...);
for await (const frame of rendered) {
   yield map(frame);
}
return map(await rendered);
```

all should be equivalent. (Not super ergonomic, I know 😦)

Reply:
There was a `JSON.parse` missing on the arguments. Should get fixed.

Reply:
hehehe yeah that first one is the one I'd reach for in terms of concision but damn would it be hard for an uninformed dev to come up with that on their own.

Reply:
I assume I can resolve this thread?